cmd/promtool/backfill_test.go

```diff
@@ -0,0 +1,115 @@
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"testing"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/prometheus/prometheus/tsdb/importer/openmetrics"
+	"github.com/prometheus/prometheus/tsdb/tsdbutil"
+	"github.com/prometheus/prometheus/util/testutil"
+)
+
+type sample1 struct {
+	t int64
+	v float64
+}
+
+func (s sample1) T() int64   { return s.t }
+func (s sample1) V() float64 { return s.v }
+
+func createTemporaryOpenmetricsFile(t *testing.T, omFile string, text string) error {
+	f, err := os.Create(omFile)
+	testutil.Ok(t, err)
+	_, errW := f.WriteString(text)
+	testutil.Ok(t, errW)
+	testutil.Ok(t, f.Close())
+	return nil
+}
+func testBlocks(t *testing.T, blockpath string, mint, maxt int64, expectedSeries map[string][]tsdbutil.Sample) {
+	db, err := tsdb.OpenDBReadOnly(blockpath, nil)
+	testutil.Ok(t, err)
+	q, err := db.Querier(context.TODO(), mint, maxt)
+	testutil.Ok(t, err)
+	ss := q.Select(false, nil, labels.MustNewMatcher(labels.MatchRegexp, "", ".*"))
+	for ss.Next() {
+		series := ss.At()
+		testutil.Equals(t, expectedSeries, series)
+	}
+	testutil.Ok(t, db.Close())
+}
+func TestBackfill(t *testing.T) {
+	tests := []struct {
+		ToParse  string
+		IsOk     bool
+		Expected struct {
+			MinTime int64
+			MaxTime int64
+			Series  map[string][]tsdbutil.Sample
+		}
+	}{
+		{
+			ToParse: `# HELP http_requests_total The total number of HTTP requests.
+# TYPE http_requests_total counter
+http_requests_total{code="200"} 1021 1565133713989
+http_requests_total{code="400"} 1 1565133713990
+# EOF
+`,
+			IsOk: true,
+			Expected: struct {
+				MinTime int64
+				MaxTime int64
+				Series  map[string][]tsdbutil.Sample
+			}{
+				MinTime: 1565133713989000,
+				MaxTime: 1565133713990000,
+				Series:  map[string][]tsdbutil.Sample{"{http_requests_total=\"200\"}": []tsdbutil.Sample{sample1{1565133713989, 1021}}, "{http_requests_total=\"400\"}": []tsdbutil.Sample{sample1{1565133713990, 1}}},
+			},
+		},
+	}
+	for _, test := range tests {
+		omFile := "backfill_test.om"
+		testutil.Ok(t, createTemporaryOpenmetricsFile(t, omFile, test.ToParse))
+		input, errOpen := os.Open(omFile)
+		testutil.Ok(t, errOpen)
+		outputDir := "./tmpDir"
+		testutil.Ok(t, backfill(input, outputDir))
+		_, errReset := input.Seek(0, 0)
+		testutil.Ok(t, errReset)
+		p := openmetrics.NewParser(input)
+		maxt, mint, errTs := getMinAndMaxTimestamps(p)
+		testutil.Equals(t, test.Expected.MinTime, mint)
+		testutil.Equals(t, test.Expected.MaxTime, maxt)
+		testutil.Ok(t, input.Close())
+		if test.IsOk {
+			blocks, _ := ioutil.ReadDir(outputDir)
+			for _, block := range blocks {
+				blockpath := filepath.Join(outputDir, block.Name())
+				testutil.Ok(t, os.MkdirAll(path.Join(blockpath, "wal"), 0777))
+				testBlocks(t, blockpath, mint, maxt, test.Expected.Series)
+			}
+		} else {
+			testutil.NotOk(t, errTs)
+		}
+		testutil.Ok(t, os.RemoveAll("backfill_test.om"))
+		testutil.Ok(t, os.RemoveAll(outputDir))
```

Comment:
Is it possible to do those cleanups in a `defer`'d call so that they also happen if the test panics or aborts before these lines are reached?

Reply:
yes, made the changes